### ISSUE TITLE
react-redux error with optional values

### ISSUE THREAD

**Issue by johnryan** [flow-typed/flow-typed#1997]
Since the recent redux changes i've been seeing some problems when using connect with optional values. See:

https://flow.org/try/#0PTAEAEDMBsHsHcBQBLAtgB1gJwC6hwJ7oCmoA3qAMKwawB2xdOAKkcQDSgCi0xqjOAApZY6AM6gAvqEgjUoAORZiAQwDGOBQG4UtXKABUoFRIBKqjTLmLl6zTsQATYmugrl+NqACyK9AGUcFRxiZlhhUTEAHlN-QQAuUAB5ACMAKxccTn9mAEFmLkTUjI0APlAAXlAACjEgkMSc-K4ASkry2MEdZ1d3UmIAD0x9SABXOg1kelA1egYNKOpURKXMBiZWEiiDUs5OovTM7LyCg5KcUurEUFBUP0Dg0PCRcUTfAPqniPEYuOPm0qIFqJaqzWjrHArGhtCrlVb0ASbYhRAAkABFkJBIFEeHwBN9oktdqBOqVSg5EIQSKACZVyKIcFM6CpoAA1FmjYgAfkSdSwyDoAHNJDpemIJEtQIMQnRHGYLDgAHTwiFRAnlMjXUDKWXELDVNqam43ZQ4UZYOigKKOZAAN1KZBwAAtkGJFegXm6GUyWezoJzJFFgDb7TobpJEBHELM6HVbvdPmFaVVap8YRqtabzZajcbvfRfRziIkAETMYh1EtaiMixDR+hx6h0eYhRx0mMt6p3D6PJOelrVJYtByLOaZYht4DkxBAA

Basically, when a type is declared as optional i.e.:
```
type Props = {
  myOptionalType?: string
}
```

Flow complains if i set it to a string using connect. My expectation would be that the example above should be valid since the intention is that `myOptionalType` will be (undefined| string).

**Comment by AndrewSouthpaw** [comment]
IIRC, you can explicitly type the return of `mapStateToProps` [in this case](https://flow.org/try/#0PTAEAEDMBsHsHcBQBLAtgB1gJwC6hwJ7oCmoA3qAMKwawB2xdOAKkcQDSgCi0xqjOAApZY6AM6gAvqEgjUoAORZiAQwDGOBQG4UtXKABUoFRIBKqjTLmLl6zTsQATYmugrl+NqACyK9AGUcFRxiZlhhUTEAHlN-QQAuUAB5ACMAKxccTn9mAEFmLkTUjI0APlAAXlAACjEgkMSc-K4ASkry2MEdZ1d3UmIAD0x9SABXOg1kelA1egYNKOpURKXMBiZWEiiDUs5OovTM7LyCg5KcUurEUFBUP0Dg0PCRcUTfAPqniPEYuOPm0qIFqJaqzWjrHArGhtCrlVb0ASbYhRAAkABFkJBIFEeHwBN9oktdqBOqVSg5EIQSKACZVyKIcFM6CpoAA1FmjYgAfkSdSwyDoAHNJDpemIJEtQIMQnRHGYLDgAHTwiFRAnlMjXUDKWXELDVNqam43ZQ4UZYOigKKOZAAN1KZBwAAtkGJFegXm6GUyWezoJzJFFgDb7TobpJEBHELM6HVbvdPmFaVVap9gTTPe1yDcQFaALQFgta03my1G43e+i+jnERIAImYxDqda1EZFiGj9Dj1Do8xCjjpMb71TuH0eSc9LWqSxaDkWc0yxAHwHJiCAA). It's not a bug with the libdef, it's an issue with Flow.

@villesau should also weigh in, in case I got that wrong. 😉

**Comment by villesau** [comment]
@AndrewSouthpaw @johnryan yes. At least to me it looks very much like some sort of inference bug, as it should behave exactly like [this](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAKA3nADgFwEs4A7AQxgDUKBXAUwC4wByAFToGd9mBfJnAsXJVadAPxMuAJ0IkA5jwCUQA)

I might of course be wrong (and glad to be proven so!), but I can't come up with an idea how to fix that by changing libdefs.

**Comment by goodmind** [comment]
I think this is related to defaultProps somehow
